### PR TITLE
Fix a bug caused by too long return expressions

### LINKED ISSUE #1252: kotlinpoet will fail if it returns an expression that is too long
When the generated `return someExpression()` expression is too long (over 100), it will cause kotlinpoet to automatically insert a newline and the IDE check will pop up a `[RETURN_TYPE_MISMATCH]` error and the compilation will fail.

And when I tried to change generate `columnLimit`, I found it very difficult to change it because `CodeWriter` is an internal class.

image

image

My SymbolProcessor and FileSpec generation code is as follows, which simply generates a long return expression:

```kotlin
@AutoService(SymbolProcessorProvider::class)
class SimpleKspProcessor : SymbolProcessorProvider {
    override fun create(environment: SymbolProcessorEnvironment): SymbolProcessor {
        return object : SymbolProcessor {
            override fun process(resolver: Resolver): List<KSAnnotated> {
                val allSeq = resolver.getSymbolsWithAnnotation(Anno::class.java.canonicalName)
                val annotated = allSeq.firstOrNull() ?: return emptyList()
                annotated.accept(object : KSVisitorVoid() {
                    override fun visitFunctionDeclaration(function: KSFunctionDeclaration, data: Unit) {
                        val fileSpec = longReturn(function.packageName.asString())
                        runCatching {
                            val file = environment.codeGenerator.createNewFile(Dependencies.ALL_FILES, fileSpec.packageName, fileSpec.name)
                            OutputStreamWriter(file, StandardCharsets.UTF_8).use(fileSpec::writeTo)
                        }
                    }
                }, Unit)
                return listOf(annotated)
            }
        }
    }
}

fun longReturn(packageName: String): FileSpec {
    val myLongClass = "Qwertyuiopasdfghjklzxcvbnmqwertyuiopasdfghjklzxcvbnmqwertyuiopasdfghjklzxcvbnmqwertyuiopasdfghjklzxcvbnm"
    val simpleClass = TypeSpec.classBuilder("KspGen")
    simpleClass.addFunction(
        FunSpec.builder("ccc")
            .addModifiers(KModifier.PUBLIC)
            .returns(ClassName.bestGuess("org.example.$myLongClass"))
            .addStatement("val a = 888")
            .addStatement("return $myLongClass()")
            .build()
    )
    return FileSpec.builder(packageName, "KspGen")
        .addType(simpleClass.build())
        .build()
}
```

---
versions:
- kotlin: 1.6.20
- kotlinpoet: 1.11.0
- ksp: 1.6.20-1.0.5

My test code is here: [zsqw123/KotlinPoetIssue](https://github.com/zsqw123/KotlinPoetIssue)

### PR DESCRIPTION
**Pull request by zsqw123**
Fix #1252 @Egorand

### CODE REVIEW

File: kotlinpoet/src/main/java/com/squareup/kotlinpoet/CodeBlock.kt [hunk 1]
```diff
@@ -270,7 +270,11 @@ public class CodeBlock private constructor(
         if (format[p] != '%') {
           var nextP = format.nextPotentialPlaceholderPosition(startIndex = p + 1)
           if (nextP == -1) nextP = format.length
-          formatParts += format.substring(p, nextP)
+          var formatPart = format.substring(p, nextP)
```
Comment (Egorand) [inline kotlinpoet/src/main/java/com/squareup/kotlinpoet/CodeBlock.kt:273]: I think this is too generic, we should probably only fixup function bodies. I believe this code belongs somewhere in `FunSpec`, likely `emit()`.

File: kotlinpoet/src/main/java/com/squareup/kotlinpoet/CodeBlock.kt [hunk 2]
```diff
@@ -458,6 +458,8 @@ public class CodeBlock private constructor(
     private const val TYPE_NAME = 2
     private val NO_ARG_PLACEHOLDERS = setOf("⇥", "⇤", "«", "»")
     internal val EMPTY = CodeBlock(emptyList(), emptyList())
+    internal const val RETURN_WITH_SPACE = "return "
```
Comment (Egorand) [inline kotlinpoet/src/main/java/com/squareup/kotlinpoet/CodeBlock.kt:461]: Revert this change?

File: kotlinpoet/src/test/java/com/squareup/kotlinpoet/FunSpecTest.kt [hunk 3]
```diff
@@ -206,6 +206,62 @@ class FunSpecTest {
...
+      """
+      |public fun foo(): kotlin.String {
+      |  val placeholder = 1
+      |  return "Loooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooong"
+      |}
+      |""".trimMargin()
+    )
+  }
+
+  @Test fun returnsLongExpressions() {
```
Comment (Egorand) [inline kotlinpoet/src/test/java/com/squareup/kotlinpoet/FunSpecTest.kt:234]: This test does not seem to add any value over the previous one, there should be existing tests that would catch a regression for a case where the substitution is not needed. I suggest deleting this test.

File: kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt [hunk 4]
```diff
@@ -245,6 +245,25 @@ public class FunSpec private constructor(
     return null
   }
 
+  private fun CodeBlock.returnsWithoutLinebreak(): CodeBlock {
+    val originCodeBlockBuilder = toBuilder()
+    val returnWithSpace = CodeBlock.RETURN_WITH_SPACE
+    originCodeBlockBuilder.formatParts.clear()
+    formatParts.mapTo(originCodeBlockBuilder.formatParts) { formatPart ->
```
Comment (Egorand) [inline kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt:252]: TBH I have a hard time understanding what this block does, can it be simplified? Essentially what we want to do is (in pseudocode):

```
if (body.startsWith(RETURN_EXPRESSION_BODY_PREFIX_SPACE)) {
  body = body.replace(RETURN_EXPRESSION_BODY_PREFIX_SPACE, RETURN_EXPRESSION_BODY_PREFIX_NBSP)
}
```

Helper methods inside `CodeBlock`, such as `withoutPrefix()`, `trim()`, `replaceAll()` can help implement that.

File: kotlinpoet/src/test/java/com/squareup/kotlinpoet/FunSpecTest.kt [hunk 5]
```diff
@@ -206,6 +206,62 @@ class FunSpecTest {
...
   }
 
+  @Test fun returnsLongExpression() {
+    val funSpec = FunSpec.builder("foo")
+      .returns(String::class)
+      .addStatement("val placeholder = 1")
+      .addStatement("return     \"Loooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooong\"")
+      .build()
+    val sb = StringBuilder()
+    CodeWriter(sb).use {
```
[the remark this answers is not in the document]
Comment (zsqw123) [reply kotlinpoet/src/test/java/com/squareup/kotlinpoet/FunSpecTest.kt:216]: The `toString` metho's columnLimit is Integer.MAX_VALUE, This does not reproduce my case

File: kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt (hunk 4 above)
Comment (JakeWharton) [reply kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt:252]: We don't need to support trimming. If the person is generating code which has leading spaces or multiple spaces between the actual return expression then they can fix that.

File: kotlinpoet/src/test/java/com/squareup/kotlinpoet/FunSpecTest.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (Egorand) [reply kotlinpoet/src/test/java/com/squareup/kotlinpoet/FunSpecTest.kt]: Ah, good point. Might be worth adding a short comment to the code explaining why `CodeWriter` is used.

File: kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt [hunk 6]
```diff
@@ -245,6 +245,19 @@ public class FunSpec private constructor(
     return null
   }
 
+  private fun CodeBlock.returnsWithoutLinebreak(): CodeBlock {
+    val originCodeBlockBuilder = toBuilder()
+    val returnWithSpace = "return "
+    val returnWithNbsp = "return·"
```
Comment (Egorand) [inline kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt:251]:
```suggestion
    val returnWithSpace = RETURN_EXPRESSION_BODY_PREFIX_SPACE
    val returnWithNbsp = RETURN_EXPRESSION_BODY_PREFIX_NBSP
```

Comment (Egorand) [reply kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt:251]: And I would actually inline these variables.

File: kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt [hunk 7]
```diff
@@ -245,6 +245,19 @@ public class FunSpec private constructor(
     return null
   }
 
+  private fun CodeBlock.returnsWithoutLinebreak(): CodeBlock {
+    val originCodeBlockBuilder = toBuilder()
```
Comment (Egorand) [inline kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt:249]: Is there a way to optimize the code to only call `toBuilder()` if there is something to replace?

File: kotlinpoet/src/test/java/com/squareup/kotlinpoet/FunSpecTest.kt [hunk 8]
```diff
@@ -206,6 +206,31 @@ class FunSpecTest {
     )
   }
 
+  @Test fun returnsLongExpression() {
+    val funSpec = FunSpec.builder("foo")
+      .returns(String::class)
+      .addStatement("val placeholder = 1")
+      .addStatement("return \"Loooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooong\"")
```
Comment (Egorand) [inline kotlinpoet/src/test/java/com/squareup/kotlinpoet/FunSpecTest.kt:213]: This is the preferred form, should not affect the output:
```suggestion
      .addStatement("return %S", "Loooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooong")
```

File: kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt [hunk 9]
```diff
@@ -245,6 +245,19 @@ public class FunSpec private constructor(
...
+    val returnWithSpace = "return "
+    val returnWithNbsp = "return·"
+    originCodeBlockBuilder.formatParts.clear()
+    formatParts.mapTo(originCodeBlockBuilder.formatParts) { formatPart ->
+      if (formatPart.startsWith(returnWithSpace)) {
+        formatPart.replaceFirst(returnWithSpace, returnWithNbsp)
+      } else formatPart
```
Comment (Egorand) [inline kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt:256]: nit: would be nice to use braces with the `else` statement too

**Review by Egorand — APPROVED** [review]
Looks great, thanks! I just added a few suggestions on the style.

Also, please sign our [CLA](https://spreadsheets.google.com/spreadsheet/viewform?formkey=dDViT2xzUHAwRkI3X3k5Z0lQM091OGc6MQ&ndplr=1) before I can merge your PR.

File: kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt (hunk 7 above)
Comment (zsqw123) [reply kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt:249]: I think so, but I'm ready to go to bed now. I'll change it tomorrow morning (I'm at UTC+8 time zone).

Comment (zsqw123) [reply kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt]: I've changed it so that the `toBuilder` is now called on demand

Comment (zsqw123) [reply kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt]: However, many functions actually have a return method, this optimization may not have a significant effect

Comment (Egorand) [reply kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt]: Yeah, good point, the code looks good though, so let's leave it as is!

**Comment by Egorand** [comment]
Merging, thanks again!

**Pull request merged**